lib/matplotlib/lines.py

```diff
@@ -89,24 +89,24 @@ def segment_hits(cx, cy, x, y, radius):
     # Only look at line segments whose nearest point to C on the line
     # lies within the segment.
     dx, dy = x[1:] - xr, y[1:] - yr
-    Lnorm_sq = dx ** 2 + dy ** 2  # Possibly want to eliminate Lnorm==0
-    u = ((cx - xr) * dx + (cy - yr) * dy) / Lnorm_sq
-    candidates = (u >= 0) & (u <= 1)
+    u = (cx - xr) * dx + (cy - yr) * dy
+    candidates = (0 <= u) & (u <= dx ** 2 + dy ** 2)
 
     # Note that there is a little area near one side of each point
     # which will be near neither segment, and another which will
     # be near both, depending on the angle of the lines.  The
     # following radius test eliminates these ambiguities.
-    point_hits = (cx - x) ** 2 + (cy - y) ** 2 <= radius ** 2
```

Comment:
Do we have an estimate how many elements this typically works on?

`np.hypot` scales worse than the squares, likely because taking the root is expensive.

![grafik](https://user-images.githubusercontent.com/2836374/80797481-3ba73c80-8ba2-11ea-8f7f-14310077ac22.png)


~~~
import numpy as np
import matplotlib.pyplot as plt

ns = [10, 100, 200, 400,600, 1000]
v1 = []
v2 = []

for n in ns:
    a = np.linspace(0, 1, n)
    b = 2*a
    a0 = 0.1
    b0 = 0.2
    r = 1.2
    o1 = %timeit -o (a0 - a)**2 + (b0 - b)**2 < r**2
    o2 = %timeit -o np.hypot(a0 - a, b0 - b) < r
    v1.append(o1)
    v2.append(o2)
    
plt.plot(ns, [v.average *1e6 for v in v1], 'o--', label='squares')
plt.plot(ns, [v.average *1e6 for v in v2], 'o--', label='np.hypot')
plt.legend()
plt.ylabel('duration [µs]')
plt.xlabel('array size')
~~~

Reply:
It is indeed slower, but on my machine the scaling is not so bad.  Moreover skipping the call to errstate() is also quite helpful, as contextmanagers are slow (relatively speaking):
```
import numpy as np 
import matplotlib.pyplot as plt 
 
ns = [10, 100, 200, 400, 600, 1000] 
v1 = [] 
v2 = [] 
v3 = [] 
 
for n in ns: 
    a = np.linspace(0, 1, n) 
    b = 2*a 
    a0 = 0.1 
    b0 = 0.2 
    r = 1.2 
    o1 = %timeit -o (a0 - a)**2 + (b0 - b)**2 < r**2 
    o2 = %timeit -o with np.errstate(all="ignore"): (a0 - a)**2 + (b0 - b)**2 < r**2 
    o3 = %timeit -o np.hypot(a0 - a, b0 - b) < r 
    v1.append(o1) 
    v2.append(o2) 
    v3.append(o3) 
 
plt.plot(ns, [v.average *1e6 for v in v1], 'o--', label='squares') 
plt.plot(ns, [v.average *1e6 for v in v2], 'o--', label='squares+errstate') 
plt.plot(ns, [v.average *1e6 for v in v3], 'o--', label='np.hypot') 
plt.legend() 
plt.ylabel('duration [µs]') 
plt.xlabel('array size')
```
![test](https://user-images.githubusercontent.com/1322974/80802285-3a304100-8baf-11ea-8867-3e4126747365.png)

(I know there's *two* calls to hypot vs a single errstate(), so I guess the real threshold where hypot becomes slower would be around 500 points.)

Reply:
Ok, let's assume we rarely deal with polygons with thousands of segments, so timing is not too different.